File: leetcode-python-solutions/longest-common-prefix.py
def longest_common_prefix(strs):
    if len(strs) == 0:
        return ""
    # sort array first
    strs.sort(key=len, reverse = False)
    print(strs)
    common_prefix = strs[0]
    for i in range(1, len(strs)):
        for j in range(len(common_prefix)):
            if strs[i][j] != common_prefix[j]:
                common_prefix = common_prefix[0:j]
                break
    print(common_prefix)
    return common_prefix


longest_common_prefix(['flower', 'flow', 'flight'])

File: leetcode/leetcode-js-solutions/longest-common-prefix/longestCommonPrefix.js
const longestCommonPrefix = (strs) => {
  if (strs.length === 0) {
    return ""; 
  }
  // Sort the array first so that the common prefix is initalised as a word that has a length larger than or equal to next words length. We are also limited by our shortest word anyway so its a good place to start. 
  let sortedStrs = strs.sort((a,b) => {return a.length-b.length}); 
  // note this has sorted array in place. original array (strs) is lost unless you use spread. we dont care about original strs though.  
  console.log(sortedStrs); 
  // initalise first word as common prefix and check against it. If at any point there is no match, slice the array. 
  let commonPrefix = sortedStrs[0];  
  // loop through array of strings
  for (let i=1; i<sortedStrs.length; i++) {
    // loop through a string
    for (let j=0; j<commonPrefix.length; j++) {
      if (sortedStrs[i][j] != commonPrefix[j]) {
        commonPrefix = commonPrefix.substring(0,j); 
        // we have to break here - otherwise we'll index for a value of j that doesnt exist in common prefix anymore since its been sliced. Breaking the inner j loop will move onto the next iteration of the i loop - checking the next word.  
        break
      }
    }
    
  }
  console.log(commonPrefix); 
  return commonPrefix;
}

longestCommonPrefix(["ab","a"])